controllers/hls.go

```diff
@@ -0,0 +1,66 @@
+package controllers
+
+import (
+	"io/ioutil"
+	"net/http"
+	"path"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/owncast/owncast/config"
+	"github.com/owncast/owncast/core"
+	"github.com/owncast/owncast/core/data"
+	"github.com/owncast/owncast/router/middleware"
+	"github.com/owncast/owncast/utils"
+	log "github.com/sirupsen/logrus"
+)
+
+// HandleHLSRequest will manage all requests to HLS content.
+func HandleHLSRequest(w http.ResponseWriter, r *http.Request) {
+	// Sanity check to limit requests to HLS file types.
+	if filepath.Ext(r.URL.Path) != ".m3u8" && filepath.Ext(r.URL.Path) != ".ts" {
+		w.WriteHeader(http.StatusNotFound)
+		return
+	}
+
+	requestedPath := r.URL.Path
+	relativePath := strings.Replace(requestedPath, "/hls/", "", 1)
+	fullPath := filepath.Join(config.HLSStoragePath, relativePath)
+
+	// If using external storage then only allow requests for the
+	// master playlist at stream.m3u8, no variants or segments.
+	if data.GetS3Config().Enabled && relativePath != "stream.m3u8" {
+		w.WriteHeader(http.StatusNotFound)
+		return
+	}
+
+	// Handle playlists
+	if path.Ext(r.URL.Path) == ".m3u8" {
+		// Playlists should never be cached.
+		middleware.DisableCache(w)
+
+		// Use this as an opportunity to mark this viewer as active.
+		id := utils.GenerateClientIDFromRequest(r)
+		core.SetViewerIdActive(id)
+	} else {
+		cacheTime := utils.GetCacheDurationSecondsForPath(relativePath)
+		w.Header().Set("Cache-Control", "public, max-age="+strconv.Itoa(cacheTime))
+	}
+
+	writeHLSResponse(fullPath, w)
+}
+
+func writeHLSResponse(filePath string, w http.ResponseWriter) {
```

Comment:
There are things missing by throwing out `http.FileServer` and just using `io.Copy`, like 404 Not Found, Range Requests, If-Unmodified-Since, or Mime-Types. Id suggest to use [`http.ServeFile`](https://pkg.go.dev/net/http#ServeFile) or [`http.ServeContent`](https://pkg.go.dev/net/http#ServeContent) which deals with all of the above.

Reply:
Good thoughts!  No reason not to use them.